### ISSUE TITLE
Canvas paint called on every MouseRegion entry

### ISSUE THREAD

**Issue by akkinenirajesh** [flutter/flutter#49298]
When ever mouse is moving on any MouseRegions/Buttons, my canvas is being repainted. There is no relation or proximity between those Widgets and Canvas needed.

On v1.13.6 Flutter Web

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          children: <Widget>[
            Container(child:
            RaisedButton(
              child: Text('Mouse over me'),
              onPressed: () {},
            ),
            margin: EdgeInsets.all(50),
            ),
            Container(child: CustomPaint(
              painter: MyPainter(),
            ), width: 200, height: 200,),
          ],
        ),
      ),
    );
  }
}
int count = 0;
class MyPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawRect(Offset.zero & size, Paint()..color = Colors.lightBlue);
    print('Paint is being called ${count++}');
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
}

```

I expect my painter not to be called on events which are not even related. 

But as I move mouse over the button in above code, painter is being called. 

I even gave enough margin between them so that the slash also is not intersecting with canvas.

**Comment by dnfield** [comment]
@dkwingsmt - does this seemright to you?

I can reproduce this on web and on macOS desktop.  I'm surprised the repaint is happening when mousing over the button.

@akkinenirajesh thanks for the great example - I've slightly modified it to make the logs and visuals more obvious.  LMK if I changed too much.

**Comment by dnfield** [comment]
(overriding `hitTest` doesn't help here either)

**Comment by dkwingsmt** [comment]
Doesn't look intended at first look. I'll take a look later.

**Comment by ferhatb** [comment]
@dkwingsmt : Do you have ETA on fix for mouse region layers ?

**Comment by dkwingsmt** [comment]
@dnfield @akkinenirajesh First, it's unlikely to be related with `MouseRegion` or enter/exit events, since repaint also happens when you click the button. 
If I understand correctly, it's the probably the issue of repaint boundary. When something is repainted, everything on the same layer needs to be repainted. Neither `CustomPaint` nor `MouseRegion` comes with a repaint boundary. By wrapping the `CustomPaint` with a `RepaintBoundary` widget, it is no longer repainted during enter/exit.

**Comment by dnfield** [comment]
Should we consider either automatically or more easily adding repaint boundaries around mouse regions?  Perhaps only if there's a mouse available (is that possible)?

**Comment by dnfield** [comment]
Talked offline with @goderbauer about this.

Adding more layers will actually hurt things on web from what I understand. So making all mouse regions contain a layer to avoid this would hurt web (and other platforms even if just to a lesser extent).

This seems like a case where the best solution is just wrap your expensive to repaint item (the CustomPainter) in a RepaintBoundary.

**Comment by akkinenirajesh** [comment]
Web and Desktop UI do usually have lot of mouse hover effects. And seems like MouseRegion is the only way to achieve that. Given that scenario, I think this seems to be fundamental problem.

There has to be better way handling Hover without repainting everything else on the screen. Because that total screen would be too big.

I don't know much about the internal details of flutter painting algorithm, How is it different than say Win32? Where we repaint one window any at time. May be I am generalizing too much, I apologise for that.

I am actually trying to build complete IDE in flutter that has lot of controls (trees, canvas, forms and lot more). Imagine VSCode + Sketch(or Figma). Wondering if I am expecting too much from Flutter. :-)

**Comment by dkwingsmt** [comment]
It is related to the rendering pipeline of Flutter, which is briefly discussed in talks like https://youtu.be/UUfXWzp0-DU . I’m not the expert on it but here’s my 2 cents.

In this particular issue we’ve been talking about the painting stage. Simply put, every object paints on a layer, which you can think of as a layer in Photoshop, and then all layers are composited (stacked) together. The catch is, a layer is either not repainted, or fully repainted. So if you have something that changes frequently, and independently from the other part of the screen, you make it a separate layer. It can be done by wrapping that part of the widget tree in a RepaintBoundary.
```dart
RepaintBoundary(
  child: CustomPaint()
)
```

However, as you would imagine, having too many layers also impacts performance. The impact by number of layers is more tolerable on desktop or mobile, but unfortunately significant on Web, which we definitely want to improve.
The bottom line is, insert RepaintBoundary in a reasonable way, and profile your performance.

**Comment by dkwingsmt** [comment]
I'm gonna close it as it is working as intended. Feel free to comment or reopen if necessary.
